### PR TITLE
test(cypress): add Connector Onboarding testcases

### PR DESCRIPTION
**Pull request by Gnanasundari24**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [x] CI/CD

## Description

Adds Cypress test coverage for the three Connector Onboarding Config endpoints:

- `POST /connector_onboarding/action_url` — retrieves the PayPal OAuth onboarding URL; expects `422 IR_06` in test environments where PayPal credentials are not configured.
- `POST /connector_onboarding/sync` — synchronises onboarding status with PayPal; expects `400` in test environments where no real PayPal integration exists.
- `POST /connector_onboarding/reset_tracking_id` — resets the onboarding tracking ID in the `configs` table; returns `200` with empty body.

### Additional Changes

- `cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js` — new spec covering happy-path, credential tests, negative cases, and edge-case scenarios (9 tests total).
- `cypress-tests/cypress/e2e/configs/Payment/Utilities/ConnectorOnboarding.js` — fixture data for each endpoint, reflecting actual responses in the test environment.
- `cypress-tests/cypress/support/commands.js` — added `connectorOnboardingActionUrl`, `connectorOnboardingSync`, and `connectorOnboardingResetTrackingId` Cypress commands.

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

The `configs` table stores per-connector onboarding state (`onboarding_{connector}_{mca_id}`). The three connector-onboarding routes were previously untested in the Cypress regression suite, leaving a gap for the PayPal onboarding flow. This PR closes that gap by adding test coverage for all three endpoints.

The tests are written in the `Misc/` test category because the endpoints use `JWTAuth` (`MerchantAccountWrite` permission) rather than the standard admin API-key auth used by payment flows. The `before` hook performs a full JWT bootstrap — user signup, 2FA skip via `/user/2fa/terminate`, and a real PayPal connector creation — so every test runs against a live MCA ID.

Related: QA pipeline issue A-73 (internal).

## How did you test it?

The spec was executed against the Hyperswitch dev server. All 9 tests pass.

```
RUNNER_RESULT:
  Connector: paypal (connector-onboarding endpoints only)
  SpecFile: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js
  TotalTests: 9
  Passed: 9
  Failed: 0
  Skipped: 0
  OverallStatus: PASS

  Test breakdown:
    Happy Path Tests
      ✓ should reset connector onboarding tracking ID
      ✓ should reset tracking ID a second time
    PayPal Credential Tests (IR_06 in dev env)
      ✓ action_url returns IR_06 when PayPal credentials are not configured (expected)
      ✓ sync returns error when no PayPal connector integration exists (expected)
    Negative Tests
      ✓ should fail action_url with non-existent MCA (404)
      ✓ should fail reset_tracking_id with non-existent MCA (404)
      ✓ should fail action_url with invalid connector type (422)
    Edge Case Tests
      ✓ should handle multiple sequential reset_tracking_id calls
      ✓ should handle action_url with special characters in return_url

  Notes:
    - action_url and sync return non-200 responses in the test environment because
      PayPal client_id/client_secret are not configured in [connector_onboarding.paypal].
      These are expected failure modes explicitly tested as part of coverage.
    - reset_tracking_id writes to the configs table and does not call PayPal APIs,
      so it returns 200 reliably in all environments.
```

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [x] I added unit tests for my changes where possible

### CODE REVIEW

**Review by XyneSpaces — COMMENTED** [review]
## Automated Review Summary

**PR:** test(cypress): add Connector Onboarding testcases
**Reviewer:** XyneSpaces
**Classification:** CI/CD - Cypress Test Coverage

---

### Findings Overview

| Severity | Count | Issues |
|----------|-------|--------|
| :warning: Warning | 1 | Inconsistent error body structure in negative test |
| :bulb: Suggestion | 2 | Missing explicit status check; assertion ordering |

---

### :white_check_mark: Positive Notes

- **Comprehensive test coverage** — 9 tests covering happy path, credential tests, negative cases, and edge cases
- **Proper JWT bootstrap** — Uses full user signup flow with 2FA termination for JWTAuth endpoints
- **Good state management** — Uses `globalState` to pass MCA ID between setup and tests
- **Clear test organization** — Tests grouped into logical contexts (Happy Path, PayPal Credential Tests, Negative Tests, Edge Cases)
- **Test environment awareness** — Correctly documents that PayPal endpoints return IR_06 in test environments without PayPal credentials
- **x-request-id logging** — All commands call `logRequestId` for observability

---

### :warning: Detailed Findings

**1. Inconsistent error body structure in negative test**

The `should fail action_url with invalid connector type` test expects a 422 response but uses a generic `{ error: {} }` body structure, which doesn't match the actual API error format used elsewhere.

**Fix:** Either use the fixture from `connectorDetails.ConnectorOnboarding.ActionUrl` with an overridden connector type, or define the expected error structure explicitly.

---

### :bulb: Suggestions

**2. Missing explicit status code assertion in `connectorOnboardingResetTrackingId`**

When `response.status === 200`, the assertion `expect(response.status).to.equal(200)` is redundant since we already verified status equality. Consider adding a check for the empty body: `expect(response.body).to.be.empty`.

**3. `Sync` command accepts either 200 OR the expected status**

The condition allows 200 to pass through unverified when the expected status is 200. Restructure to clearly distinguish success (200) from expected failure (400 in test env).

---

**Verdict:** Good test coverage with minor suggestions. The test fixture pattern and state management are clean and follow Cypress conventions.

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js [hunk 1]
```diff
@@ -0,0 +1,209 @@
+import State from "../../../utils/State";
```
Comment (XyneSpaces) [inline cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js:1]: :warning: **Inconsistent error body structure**

This test manually constructs an error response structure `{ error: {} }` which differs from the actual API error format used elsewhere. The test should either:
1. Use the `ActionUrl` fixture with connector override
2. Or define the full expected error structure with `type` and `code` fields

This ensures the test actually validates the error response format, not just that an error exists.

File: cypress-tests/cypress/support/commands.js [hunk 2]
```diff
@@ -10182,3 +10182,149 @@ Cypress.Commands.add(
     });
```
Comment (XyneSpaces) [inline cypress-tests/cypress/support/commands.js:10965]: :bulb: **Missing response body validation on 200**

When response.status === 200, the assertion is redundant. Consider adding a check for the empty body:
```javascript
expect(response.body).to.be.empty; // reset_tracking_id returns empty body
```

This ensures the endpoint contract is fully verified.

File: cypress-tests/cypress/support/commands.js [hunk 3]
```diff
@@ -10182,3 +10182,149 @@ Cypress.Commands.add(
...
+ * Retrieves PayPal onboarding action URL for OAuth flow
+ * Requires JWT authentication (admin API key)
+ * Stores tracking_id in globalState for subsequent calls
+ */
+Cypress.Commands.add(
+  "connectorOnboardingActionUrl",
+  (requestBody, data, globalState) => {
+    const baseUrl = globalState.get("baseUrl");
+    const userInfoToken = globalState.get("userInfoToken");
+
```
Comment (XyneSpaces) [inline cypress-tests/cypress/support/commands.js:11020]: :bulb: **Logic allows 200 to pass unverified**

The condition `response.status === 400 || response.status === expectedStatus` will evaluate to true for 200 responses when expectedStatus is 200, but then only checks for error properties which will fail.

Restructure to clearly separate success vs expected error paths.

**Review by XyneSpaces — COMMENTED** [review]
## Automated Review Summary

**PR:** test(cypress): add Connector Onboarding testcases  
**Type:** CI/CD — Cypress Test Coverage  
**Files Changed:** 3 files (JS test files)

---

### Findings Overview

| Severity | Count | Issues |
|----------|-------|--------|
| :warning: Warning | 2 | Missing cleanup in after hook; hardcoded test credentials |
| :bulb: Suggestion | 1 | Test assertions incomplete for 200 case |

---

### :white_check_mark: Positive Notes

- Comprehensive test coverage for all 3 connector onboarding endpoints (action_url, sync, reset_tracking_id)
- Good categorization of tests (Happy Path, Negative Tests, Edge Cases, PayPal Credential Tests)
- Proper JWT bootstrap flow with 2FA termination pattern
- Consistent use of `failOnStatusCode: false` for negative test cases
- Good documentation comments in commands.js explaining each endpoint

---

### :warning: Detailed Findings

**1. Missing cleanup/teardown in `after` hook**

The `after` hook only flushes global state but doesn't clean up the test resources created during setup (user, PayPal connector). Over time, this will accumulate orphaned test data in the database.

```javascript
// Line 75-77 in 00002-ConnectorOnboarding.cy.js
after("flush global state", () => {
  cy.task("setGlobalState", globalState.data);
});
```

**Fix:** Add cleanup calls to delete the PayPal connector and user:
```javascript
after("cleanup test resources", () => {
  // Delete PayPal connector
  cy.request({
    method: "DELETE",
    url: `${baseUrl}/account/${globalState.get("merchantId")}/connectors/${globalState.get("paypalConnectorId")}`,
    headers: { Authorization: `Bearer ${globalState.get("userInfoToken")}` },
    failOnStatusCode: false,
  });
  cy.task("setGlobalState", globalState.data);
});
```



---

**2. Hardcoded test credentials in connector creation**

```javascript
// Line 53-58 in 00002-ConnectorOnboarding.cy.js
connector_account_details: {
  auth_type: "BodyKey",
  api_key: "test_paypal_key",
  key1: "test_paypal_secret",
},
```

While these are test credentials, they should ideally come from environment variables or fixtures for flexibility across different test environments.

**Fix:** Move credentials to fixture file or environment variables:
```javascript
const paypalCreds = Cypress.env("PAYPAL_TEST_CREDS") || {
  api_key: "test_paypal_key",
  key1: "test_paypal_secret",
};
```



---

### :bulb: Suggestions

**3. Incomplete assertion for `action_url` 200 case**

In `commands.js`, the success case only checks that `action_url` is not null but doesn't validate it's a proper URL:

```javascript
// Line 10212-10214
expect(response.body).to.have.property("paypal");
expect(response.body.paypal).to.have.property("action_url");
expect(response.body.paypal.action_url).to.not.be.null;
```

**Fix:** Add URL format validation:
```javascript
expect(response.body.paypal.action_url).to.match(/^https:\/\/.+/);
```



---

### Pre-existing Issues

None identified.

---

**Verdict:** Clean test implementation following established Cypress patterns. Address the cleanup gap (issue 1) before merging to prevent test data accumulation.

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+describe("Connector Onboarding", () => {
+  before("setup user JWT and PayPal connector", () => {
+    cy.task("getGlobalState").then((state) => {
+      globalState = new State(state);
+
+      const baseUrl = globalState.get("baseUrl");
+      const ts = Date.now();
+
+      // 1. Signup to get TOTP token
+      cy.request({
```
Comment (Gnanasundari24) [inline cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js]: Dont add cy.request in the spec add the existing function else create it in commands.js

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+          company_name: "Test Co",
+        },
+        failOnStatusCode: false,
+      }).then((signupResp) => {
+        expect(signupResp.status).to.equal(200);
+        const totpToken = signupResp.body.token;
+        globalState.set("totpToken", totpToken);
+
+        // 2. Terminate 2FA to get full user JWT
+        cy.request({
```
Comment (Gnanasundari24) [inline cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js]: Dont add cy.request in the spec add the existing function else create it in commands.js

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+          const userInfoToken = terminateResp.body.token;
+          globalState.set("userInfoToken", userInfoToken);
+
+          // Decode JWT to extract merchant_id and profile_id
+          const payload = JSON.parse(atob(userInfoToken.split(".")[1]));
+          globalState.set("merchantId", payload.merchant_id);
+          globalState.set("profileId", payload.profile_id);
+
+          // 3. Create PayPal connector to get MCA ID
+          cy.request({
```
Comment (Gnanasundari24) [inline cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js]: Dont add cy.request in the spec add the existing function else create it in commands.js

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js (hunk 4 above)
Comment (Gnanasundari24) [reply cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js]: Fixed — all three `cy.request` calls from the `before` hook have been moved to a new `connectorOnboardingBootstrap` command in `commands.js`. The spec now just calls `cy.connectorOnboardingBootstrap(globalState)`.

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js (hunk 5 above)
Comment (Gnanasundari24) [reply cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js]: Fixed — moved to `connectorOnboardingBootstrap` in `commands.js`.

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js (hunk 6 above)
Comment (Gnanasundari24) [reply cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js]: Fixed — moved to `connectorOnboardingBootstrap` in `commands.js`.

File: cypress-tests/cypress/support/commands.js (hunk 2 above)
Comment (Gnanasundari24) [reply cypress-tests/cypress/support/commands.js:10965]: Fixed — replaced the redundant status check with `expect(response.body).to.be.empty` to properly verify the empty-body contract.

File: cypress-tests/cypress/support/commands.js (hunk 3 above)
Comment (Gnanasundari24) [reply cypress-tests/cypress/support/commands.js:11020]: Fixed — restructured to three clear branches: `200` (success body validation), `expectedStatus` (expected error path checks `error.type`), and `else` (unexpected via `defaultErrorHandler`). The `400 ||` overlap is gone.

File: cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js (hunk 1 above)
Comment (Gnanasundari24) [reply cypress-tests/cypress/e2e/spec/Misc/00002-ConnectorOnboarding.cy.js:1]: Fixed — negative tests now use `{ error: { type: "invalid_request" } }` instead of bare `{ error: {} }`. The `connectorOnboardingActionUrl` command also validates `error.type` and `error.code` from the fixture when present.
